llvm/lib/CodeGen/InterleavedAccessPass.cpp

```diff
@@ -587,6 +587,19 @@ static Value *getMask(Value *WideMask, unsigned Factor,
     }
   }
 
+  if (auto *SVI = dyn_cast<ShuffleVectorInst>(WideMask)) {
+    unsigned LeafMaskLen = LeafValueEC.getFixedValue();
+    if (SVI->isInterleave(Factor) &&
+        llvm::all_of(SVI->getShuffleMask(),
+                     [&](int Idx) { return Idx < (int)LeafMaskLen; })) {
+      auto *LeafMaskTy =
+          FixedVectorType::get(Type::getInt1Ty(SVI->getContext()), LeafMaskLen);
+      IRBuilder<> Builder(SVI);
+      return Builder.CreateExtractVector(LeafMaskTy, SVI->getOperand(0),
+                                         uint64_t(0));
```

Comment:
Why do we extract a subvector here? Should we be bailing if SVI->getOperand(0) != LeafMaskTy?

Reply:
Great example, I had to think about this one.  After thinking about it for a bit, I decided to just write the test case, and try it.  :)

You're correct, updated patch forthcoming.

Reply:
I reworked this code both to fix this issue, and to be a bit more paranoid about checking the start indices.  Added a negative test based on your example.